plugins/silero/src/vad.ts

```diff
@@ -0,0 +1,321 @@
+          const availableInferenceSamples = inferenceFrames
+            .map((x) => x.samplesPerChannel)
+            .reduce((acc, x) => acc + x, 0);
+
+          if (availableInferenceSamples < this.#model.windowSizeSamples) {
+            break; // not enough samples to run inference
+          }
+
+          const inputFrame = mergeFrames(inputFrames);
+          const inferenceFrame = mergeFrames(inferenceFrames);
+
+          // convert data to f32
+          inferenceData = Float32Array.from(
+            inferenceFrame.data.subarray(0, this.#model.windowSizeSamples),
+            (x) => x / 32767,
+          );
+
+          const p = await this.#model
+            .run(inferenceData)
+            .then((data) => this.#expFilter.apply(1, data));
+
+          const windowDuration = (this.#model.windowSizeSamples / this.#opts.sampleRate) * 1000;
+          pubCurrentSample += this.#model.windowSizeSamples;
+          pubTimestamp += windowDuration;
+          const resamplingRatio = pubSampleRate / this.#model.sampleRate;
+          const toCopy = this.#model.windowSizeSamples * resamplingRatio + inputCopyRemainingFrac;
+          const toCopyInt = Math.trunc(toCopy);
+          inputCopyRemainingFrac = toCopy - toCopyInt;
+
+          // copy the inference window to the speech buffer
+          const availableSpace = speechBuffer.length - speechBufferIndex;
+          const toCopyBuffer = Math.min(this.#model.windowSizeSamples, availableSpace);
+          if (toCopyBuffer > 0) {
+            speechBuffer.set(inputFrame.data.subarray(0, toCopyBuffer), speechBufferIndex);
+          } else if (!speechBufferMaxReached) {
+            speechBufferMaxReached = true;
+            this.#logger.warn(
+              'max_buffered_speech reached, ignoring further data for the current speech input',
+            );
+          }
+
+          const inferenceDuration = Number((process.hrtime.bigint() - startTime) / BigInt(1000000));
+          this.#extraInferenceTime = Math.max(
+            0,
+            this.#extraInferenceTime + inferenceDuration - windowDuration,
+          );
+          if (this.#extraInferenceTime > SLOW_INFERENCE_THRESHOLD) {
+            this.#logger
+              .child({ delay: this.#extraInferenceTime })
+              .warn('inference is slower than realtime');
+          }
+
+          if (pubSpeaking) {
+            pubSpeechDuration += inferenceDuration;
+          } else {
+            pubSilenceDuration += inferenceDuration;
+          }
+
+          this.queue.put({
+            type: VADEventType.INFERENCE_DONE,
+            samplesIndex: pubCurrentSample,
+            timestamp: pubTimestamp,
+            silenceDuration: pubSilenceDuration,
+            speechDuration: pubSpeechDuration,
+            probability: p,
+            inferenceDuration,
+            frames: [
+              new AudioFrame(inputFrame.data.subarray(0, toCopyInt), pubSampleRate, 1, toCopyInt),
```

Comment:
Is subarray returning a view? If so, we need to copy instead

Reply:
ah yeah subarray returns a slice of the same array.